### PR TITLE
[SPARK-15204][SQL] improve nullability inference for Aggregator

### PR DESCRIPTION
**Pull request by koertkuipers**
## What changes were proposed in this pull request?

TypedAggregateExpression sets nullable based on the schema of the outputEncoder 
## How was this patch tested?

Add test in DatasetAggregatorSuite

### CODE REVIEW

**Comment by rxin** [comment]
cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala [hunk 1]
```diff
@@ -51,7 +52,8 @@ object TypedAggregateExpression {
       bufferDeserializer,
       outputEncoder.serializer,
       outputEncoder.deserializer.dataType,
-      outputType)
+      outputType,
+      outputNullable)
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala]: I think we can just do: `!outputEncoder.flat || outputEncoder.schema.head.nullable`

Comment (koertkuipers) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala]: ok

**Comment by cloud-fan** [comment]
can we wait for https://github.com/apache/spark/pull/13553? thanks!

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala [hunk 2]
```diff
@@ -305,4 +305,13 @@ class DatasetAggregatorSuite extends QueryTest with SharedSQLContext {
     val ds = Seq(1, 2, 3).toDS()
     checkDataset(ds.select(MapTypeBufferAgg.toColumn), 1)
   }
+
+  test("spark-15204 improve nullability inference for Aggregator") {
+    val ds1 = Seq(1, 3, 2, 5).toDS()
+    assert(ds1.select(typed.sum((i: Int) => i)).schema.head.nullable === false)
+    val ds2 = Seq(AggData(1, "a"), AggData(2, "a")).toDS()
+    assert(ds2.groupByKey(_.b).agg(SeqAgg.toColumn).schema(1).nullable === true)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala]: should we use `ds.select` for all tests? And we should also test String, which is flat but nullable.

Comment (koertkuipers) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala]: the last assert with NameAgg tests String as output of the Aggregator. is that good enough?

**Comment by cloud-fan** [comment]
merging to master, thanks!

**Pull request closed without merging**